tests-l4lb: Use Helm chart from local branch

michi-covalent opened the pull request:
In addition to the upstream (i.e. v1.10, v1.11 and master branches)
checkouts, checkout also the pull request branch so that test.sh uses
the local Helm chart from the pull request. This ensures that any
Helm-related changes get reflected in the CI run.

Co-authored-by: Michi Mutsuzaki <michi@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

sayboras reviewed (approved):
:rocket:

jibi reviewed (approved):
:ship: :ship:

nbusseneau commented on .github/workflows/tests-l4lb.yaml:
```diff
@@ -159,15 +159,24 @@ jobs:
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b
+        with:
+          ref: ${{ steps.vars.outputs.sha }}
+          persist-credentials: false
```
We cannot allow `test.sh` to be modified in PRs, otherwise an attacker could set up a remote shell in the script and get access on the GitHub runner and then do things, in effect doing the same thing as editing the workflow file.

The only thing that we should allow to be taken from the PRs here is the Helm chart itself, so I would propose the following instead:
- Keep the `upstream` checkout as the main checkout (i.e. in current workdir, not a in subdir).
- Checkout PR in a subdir.
- Pass the Helm chart dir in the subdir as an argument to `test.sh`.

I think reversing the way we do the checkouts is good because then the default checkout / local dir is the safe version, and potential future editors of the workflow need not know about all the intricacies to avoid introducing security issues by mistake (because they will hit the same wall that you did already).

nbusseneau reviewed (changes requested):
Let's go, we're on the right path.

jibi committed:
tests-l4lb: Use Helm chart from local branch

In addition to the upstream (i.e. v1.10, v1.11 and master branches)
checkouts, checkout also the pull request branch so that test.sh uses
the local Helm chart from the pull request. This ensures that any
Helm-related changes get reflected in the CI run.

Co-authored-by: Michi Mutsuzaki <michi@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

tklauser commented on .github/workflows/tests-l4lb-v1.10.yaml:
```diff
@@ -183,7 +191,7 @@ jobs:
       - name: Run tests
         run: |
-          ssh default "sudo /bin/sh -c 'cd /vagrant/test/l4lb && ./test.sh ${{ github.repository_owner}} ${{ steps.vars.outputs.sha }}'"
+          ssh default "sudo /bin/sh -c 'cd /vagrant/test/l4lb && ./test.sh ${{ github.repository_owner}} ${{ steps.vars.outputs.sha }} ${{ github.workspace }}/pull-request'"
```
Don't we need to use `test.sh` from the PR checkout as well? Or otherwise we'd need to change the script in `v1.10` and `v1.11` branch to take the third argument with the helm chart directory as well.

tklauser replied:
Please disregard, I just saw https://github.com/cilium/cilium/pull/19953#discussion_r881793284. I think we still need to change `test.sh` in `v1.10` and `v1.11` to take the third argument, though.

tklauser commented on test/l4lb/test.sh:
```diff
@@ -4,6 +4,7 @@ set -eux
 
 IMG_OWNER=${1:-cilium}
 IMG_TAG=${2:-latest}
+PULL_REQUEST_CHECKOUT=${3:-latest}
```
I think we should default to `../..` for the directory here for backwards compatibility:

```suggestion
PULL_REQUEST_CHECKOUT=${3:-../..}
```

tklauser commented on test/nat46x64/test.sh:
```diff
@@ -4,6 +4,7 @@ set -eux
 
 IMG_OWNER=${1:-cilium}
 IMG_TAG=${2:-latest}
+PULL_REQUEST_CHECKOUT=${3:-latest}
```
Same here:

```suggestion
PULL_REQUEST_CHECKOUT=${3:-../..}
```

nbusseneau reviewed (approved):
Thanks.
